core/src/apps/cardano/sign_tx.py

```diff
@@ -117,13 +138,60 @@ def _validate_outputs(
         raise wire.ProcessError("Total transaction amount is out of range!")
 
 
+def _validate_certificates(certificates: List[CardanoTxCertificateType]) -> None:
+    for certificate in certificates:
+        if not is_staking_path(certificate.path):
+            raise INVALID_CERTIFICATE
+
+        if certificate.type == CardanoCertificateType.STAKE_DELEGATION:
+            if certificate.pool is None or len(certificate.pool) != POOL_HASH_SIZE:
+                raise INVALID_CERTIFICATE
+
+
+def _validate_withdrawals(withdrawals: List[CardanoTxWithdrawalType]) -> None:
+    for withdrawal in withdrawals:
+        if not is_staking_path(withdrawal.path):
+            raise INVALID_WITHDRAWAL
+
+        if not 0 <= withdrawal.amount < LOVELACE_MAX_SUPPLY:
+            raise INVALID_WITHDRAWAL
+
+
+def _validate_metadata(metadata: bytes) -> None:
+    if not metadata:
+        return
+
+    if len(metadata) > MAX_METADATA_LENGTH:
+        raise INVALID_METADATA
+
+    try:
+        # this also raises an error if there's some data remaining
+        decoded = cbor.decode(metadata)
+    except:
```

Comment:
Let's use this @gabrielKerekes - this should fix the error.

```suggestion
    except Exception:
```